src/Data/Bin/Properties.agda

```diff
@@ -1,143 +1,913 @@
+*-identityˡ x = begin
+  1B * x                 ≡⟨ sym (fromℕ∘toℕ (1B * x)) ⟩
+  fromℕ (toℕ (1B * x))   ≡⟨ cong fromℕ (toℕ*homo 1B x) ⟩
+  fromℕ (1 *ₙ n)         ≡⟨ cong fromℕ (ℕp.*-identityˡ n) ⟩
+  fromℕ n                ≡⟨ fromℕ∘toℕ x ⟩
+  x                      ∎
+  where
+  n = toℕ x
+
+*-identityʳ : RightIdentity 1B _*_
+*-identityʳ x =  trans (*-comm x 1B) (*-identityˡ x)
+
+*-identity : Identity 1B _*_
+*-identity = (*-identityˡ , *-identityʳ)
+
+2B*x≡x+x :  ∀ x → 2B * x ≡ x + x
+2B*x≡x+x x = begin
+  2B * x            ≡⟨⟩
+  (1B + 1B) * x     ≡⟨ *-distribʳ-+ x 1B 1B ⟩
+  1B * x + 1B * x   ≡⟨ cong₂ _+_ (*-identityˡ x) (*-identityˡ x) ⟩
+  x + x             ∎
+
+2x≡x+x :  ∀ x → 2* x ≡ x + x
+2x≡x+x x =  trans (2*≗2B* x) (2B*x≡x+x x)
+
+[1+x]* :  ∀ x y → (1B + x) * y ≡ y + x * y
+[1+x]* x y = begin
+  (1B + x) * y     ≡⟨ *-distribʳ-+ y 1B x ⟩
+  1B * y + x * y   ≡⟨ cong (_+ (x * y)) (*-identityˡ y) ⟩
+  y + x * y        ∎
+
+*[1+x] :  ∀ x y → y * (1B + x) ≡ y + y * x
+*[1+x] x y = begin
+  y * (1B + x)     ≡⟨ *-distribˡ-+ y 1B x ⟩
+  y * 1B + y * x   ≡⟨ cong (_+ (y * x)) (*-identityʳ y) ⟩
+  y + y * x        ∎
+
+suc* :  ∀ x y → (suc x) * y ≡ y + x * y
+suc* x y = begin
+  (suc x) * y     ≡⟨ cong (_* y) (suc≗1+ x) ⟩
+  (1B + x) * y    ≡⟨ [1+x]* x y ⟩
+  y + x * y       ∎
+
+*suc :  ∀ x y → x * (suc y) ≡ x + x * y
+*suc x y = begin
+  x * (suc y)    ≡⟨ cong (x *_) (suc≗1+ y) ⟩
+  x * (1B + y)   ≡⟨ *[1+x] y x ⟩
+  x + x * y      ∎
+
+------------------------------------------------------------------------------
+-- Multiplicative structure instances.
+-- Instances for Bin up to CommutativeSemiring.
+------------------------------------------------------------------------------
+
+*-isMagma : IsMagma _≡_ _*_
+*-isMagma = record
   { isEquivalence = isEquivalence
-  ; trans         = <-trans
-  ; compare       = <-cmp
+  ; ∙-cong        = cong₂ _*_
+  }
+
+*-magma : Magma 0ℓ 0ℓ
+*-magma = record{ isMagma = *-isMagma }
+
+*-isSemigroup : IsSemigroup _≡_ _*_
+*-isSemigroup = record
+  { isMagma = *-isMagma
+  ; assoc   = *-assoc
+  }
+
+*-semigroup : Semigroup 0ℓ 0ℓ
+*-semigroup = record{ isSemigroup = *-isSemigroup }
+
+*-1-isMonoid : IsMonoid _≡_ _*_ 1B
+*-1-isMonoid = record
+  { isSemigroup = *-isSemigroup
+  ; identity    = *-identity
+  }
+
+*-1-monoid : Monoid 0ℓ 0ℓ
+*-1-monoid = record
+  { ε        = 1B
+  ; isMonoid = *-1-isMonoid
   }
 
-<-strictTotalOrder : StrictTotalOrder _ _ _
-<-strictTotalOrder = record
-  { Carrier            = Bin
-  ; _≈_                = _≡_
-  ; _<_                = _<_
-  ; isStrictTotalOrder = <-isStrictTotalOrder
+*-1-isCommutativeMonoid :  IsCommutativeMonoid _≡_ _*_ 1B
+*-1-isCommutativeMonoid =  record
+  { isSemigroup = *-isSemigroup
+  ; identityˡ   = *-identityˡ
+  ; comm        = *-comm
   }
+
+*-1-commutativeMonoid : CommutativeMonoid 0ℓ 0ℓ
+*-1-commutativeMonoid = record{ isCommutativeMonoid = *-1-isCommutativeMonoid
+                              }
+
+isSemiringWithoutAnnihilatingZero :
```

Comment:
These records should all have `*-+-` in front of them, e.g. `*-+-isSemiringWithoutAnnihilatingZero`.

Reply:
Done